### PR TITLE
Add multiplot support and refactor plot.hpp

### PR DESCRIPTION
**Pull request by HorstBaerbel**
This, is rather big, but please take your time to review this.
* Add support for GNUplot [multiplots](http://gnuplot.sourceforge.net/docs_4.2/node203.html). This adds the class multiplot that lets you add() plots to it and then show() or save() as with plot. See the example-multiplot.cpp.
* Remove all std::ofstream and FILE* members from plot / multiplot. It is not a good idea to keep those around longer than you need them. It also made the plot class non-copyable (copying an allocated pointer / file pointer is usually no-no). The pipe is closed directly after popen() using a pclose(), which waits for the termination of the sub-process.
* Files are closed when they have been written to. The data file is reopened for every plot::draw() call and closed after writing data, which eliminates the need to flush() it. The script file is closed too after writing.
* The plot index counter has been reworked. There's now an id() function that returns the plot "number" the plot got on creation.
* GNUplot specific outputs have been extracted as functions and moved to util.hpp.
* I removed all line breaks inside functions. They make code hard to read imo and there's lots of comments for "visual line-breaks" too.
* Update README a bit.

### CODE REVIEW

File: sciplot/multiplot.hpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+#include <sciplot/specs/ticspecs.hpp>
+#include <sciplot/util.hpp>
+
+namespace sciplot
+{
+
+    using namespace internal;
+
+    /// The class used to define multipleplots in one canvas, show them in a pop-up window or save them to a file.
+    class multiplot : public plot
```
Comment (allanleal) [inline sciplot/multiplot.hpp:52]: Is inheritance here really needed?

File: sciplot/multiplot.hpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        /// Thus, to save a plot in `png` format, choose a file name with a `.png`
+        /// file extension as in `fig.png`.
+        auto save(const std::string &filename) -> void;
+
+    protected:
+        /// Plot id derived from m_counter upon construction
+        auto id() const -> std::size_t;
+
+        /// Counter of how many plot / singleplot objects have been instanciated in the application
+        static std::size_t m_counter;
```
Comment (allanleal) [inline sciplot/multiplot.hpp:90]: How do we use this `static` counter? Is it possible not to use `static`?

File: sciplot/plot.hpp [hunk 3]
```diff
@@ -42,334 +42,295 @@
 #include <sciplot/specs/ticspecs.hpp>
 #include <sciplot/util.hpp>
 
-namespace sciplot {
-
-using namespace internal;
-
-/// The class used to define the plots, show in a pop-up window and save to a file.
-class plot
+namespace sciplot
```
Comment (allanleal) [inline sciplot/plot.hpp:45]: What about no indentation inside `namespace`:

~~~cpp
namespace sciplot {
...
} // namespace sciplot

**Comment by HorstBaerbel** [comment]
About the multiplit->plot inheritance: Good catch! Definitely not needed! Wtf did it to there?!

File: sciplot/multiplot.hpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+#include <sciplot/specs/legendspecs.hpp>
+#include <sciplot/specs/linespecs.hpp>
+#include <sciplot/specs/plotspecs.hpp>
+#include <sciplot/specs/ticspecs.hpp>
+#include <sciplot/util.hpp>
+
+namespace sciplot
+{
+
+    using namespace internal;
```
[the remark this answers is not in the document]
Comment (HorstBaerbel) [reply sciplot/multiplot.hpp:49]: I'm not sure I understand your comment. I haven't given this much thought. I basically stole from plot.hpp.
Imo it is bad practice to using namespace in a header file (if find it ok in source files), so maybe we should clean this up...

File: sciplot/multiplot.hpp (hunk 2 above)
Comment (HorstBaerbel) [reply sciplot/multiplot.hpp:90]: Well, some way the plot / multiplot instances have to be counted to not overwrite files and to connect script to data files. Callers don't care about that detail, so this way it is hidden from them. We could remove the id() method and just use m_id (yagni).

File: sciplot/plot.hpp (hunk 3 above)
Comment (HorstBaerbel) [reply sciplot/plot.hpp:45]: Ok with me. I rely on auto-formatting, so I could check if I can prepare a .clang-format file that does that. Should I commit that so we can use it in the CI too?

File: sciplot/multiplot.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (HorstBaerbel) [reply sciplot/multiplot.hpp:49]: Why are classes like axislabelspecs sciplot::**internal** btw when the user uses them?!

**Comment by HorstBaerbel** [comment]
I've removed the "using namespace internal" from plot.hpp and multiplot.hpp, so users shouldn't come into contact with sciplot::internal anymore. As a library developer you have to write more, but imo it's better that the lib looks good on the user side and hides it internals.  
I've moved all the functions formatting or writing gnuplot data or running the scripts to a "gnuplot" namespace. They were prepended with "gnuplot" anyway...

**Comment by HorstBaerbel** [comment]
I know I can merge now myself (thanks for the trust btw), but I'd definitely like your opinion on changes...

File: sciplot/multiplot.hpp (hunk 2 above)
Comment (HorstBaerbel) [reply sciplot/multiplot.hpp:85]: I removed the id() function now and [made the counter private again](https://github.com/sciplot/sciplot/pull/10/commits/4d0aaf53100658a5c9649dfcb8f56a0350f14883). If we use [random temporary files](https://en.cppreference.com/w/cpp/io/c/tmpnam), we probably could get rid of it.

File: sciplot/multiplot.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (HorstBaerbel) [reply sciplot/multiplot.hpp]: This is fixed now in 9e58f1d

**Comment by HorstBaerbel** [comment]
Added a .clang-format file that can be used with the clang-format executable, but also with the Clang-format extension (xaver.clang-format) in VS Code. Turn on "Editor: Format on save" in the VS Code settings and everything is always properly formatted. If you don't like something, change the .clang-format file.
Changes (imo pretty minor):
* namespace definitions have a line break before the brace
* visibility statements (e.g. private:) are indented 2 spaces
* constructor initializer lists are indented 4 spaces
* functions longer than one line (or 80 characters?) have line breaks added

File: sciplot/plot.hpp (hunk 3 above)
Comment (HorstBaerbel) [reply sciplot/plot.hpp:45]: See 23664c8 to check if this is ok with you.

File: sciplot/multiplot.hpp (hunk 2 above)
Comment (allanleal) [reply sciplot/multiplot.hpp]: OK, now I remember the use of this static counter. The disadvantage of using random temp files is that it is not clear which generated files correspond to the sequence of plot calls. I mean, if two plots are generated, their order indication in the file name may be useful to some users, for further inspecting the corresponding generated gnuplot scripts (check my comment here in the other issue you created #12 ). In the future, we could have different options on how files are generated, name schemes, and cleaning procedure for the generated gnuplot scripts.

File: sciplot/plot.hpp (hunk 3 above)
Comment (allanleal) [reply sciplot/plot.hpp:45]: This is all fine as long as we can slightly adapt it in the future if needed! ;)

File: examples/example-multiplot.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+// sciplot includes
+#include <sciplot/sciplot.hpp>
+using namespace sciplot;
+
+int main(int argc, char **argv)
+{
+    // Create 4 different plots
+    const vec x = linspace(0.0, 5.0, 200);
+    plot p0;
```
Comment (allanleal) [inline examples/example-multiplot.cpp:34]: I think this nice example would be better presented if each block of plot commands are separated by empty lines.

~~~cpp
plot p0;
p0.draw(x, std::sin(x)).title("sin(x)");
p0.draw(x, std::cos(x)).title("cos(x)");

plot p1;
p1.draw(x, std::cos(x)).title("cos(x)");

plot p2;
p2.draw(x, std::tan(x)).title("tan(x)");

plot p3;
p3.draw(x, std::sqrt(x)).title("sqrt(x)");
~~~

File: sciplot/specs/gridspecs.hpp [hunk 6]
```diff
@@ -30,68 +30,77 @@
...
-    auto y2tics() -> gridspecsbase& { return _gridmajor("y2tics"); }
+    auto y2tics() -> gridspecsbase & { return _gridmajor("y2tics"); }
 
     /// Return a grid specification object for configuring grid lines along minor xtics.
-    auto mxtics() -> gridspecsbase& { return _gridminor("mxtics"); }
+    auto mxtics() -> gridspecsbase & { return _gridminor("mxtics"); }
 
     /// Return a grid specification object for configuring grid lines along minor ytics.
-    auto mytics() -> gridspecsbase& { return _gridminor("mytics"); }
+    auto mytics() -> gridspecsbase & { return _gridminor("mytics"); }
```
Comment (allanleal) [inline sciplot/specs/gridspecs.hpp:68]: I think it is more common to have `&` attached to the type (i.e. `gridspecsbase&`).

File: sciplot/multiplot.hpp (hunk 2 above)
Comment (HorstBaerbel) [reply sciplot/multiplot.hpp]: Let's take the discussion about this to #12

File: README.md [hunk 7]
```diff
@@ -5,49 +5,50 @@
...
+    // Draw a cosine graph putting x on the x- and cos(x) on the y-axis
     plt.draw(x, std::cos(x)).title("cos(x)").linewidth(8);
-
+    // Show the plot
     plt.show();
+    // Save the plot to a PDF file
     plt.save("figure.pdf");
 }
-~~~
+```
```
[the remark this answers is not in the document]
Comment (HorstBaerbel) [reply README.md:35]: [Googles convention ](https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace)basically says to add vertical space where it makes sense. I'm all for dense code, so as much code as possible (without being illegible) fits on one page (horizontally and vertically).
I'd suggest vertical space where there's a logical break in code flow, e.g. where you'd make a pargraph in text, but not after every line. Like:
```c++
...
// Do complicated thing A
statement();
statement();

// Now the result from A is here, now do B
statement();
statement();
...
```
Would that be ok with you? I can format plot.hpp and multiplot.hpp in that way.

File: sciplot/specs/gridspecs.hpp (hunk 6 above)
Comment (HorstBaerbel) [reply sciplot/specs/gridspecs.hpp:68]: This is somewhat of a religious thing. [Googles convention](https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions) is "use one convention per file and stick with it". Stackoverflow answers very. I don't care much, so I just set "PointerAlignment: Left". Will refomat the files.

**Comment by allanleal** [comment]
OK, not a consistent convention then! :)

**Comment by allanleal** [comment]
This is fine - vertical space when we start a new block of code as in the
example above.

**Comment by allanleal** [comment]
Sounds good!

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (HorstBaerbel) [reply README.md:41]: Fixed in d6b15f3

**Comment by allanleal** [comment]
Please request my review when you are done and I'll just have a quick look
at this and we merge this PR.

**Comment by allanleal** [comment]
Great. Thank you.

**Comment by HorstBaerbel** [comment]
Ah, sorry. I though I'd merge as I addressed all the thing you pointed out to. If there's stuff left over, we can always do this in the next PR...

**Comment by allanleal** [comment]
No worries!

**Pull request merged**